ls1intum/Artemis
# `Development`: Add exam and exercise metrics

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
In the future, Artemis should support automatic scaling based on user demand.
When more users are expected to use Artemis, Artemis should scale up and should scale down again once the increased demand is over.

We currently expose the metric `artemis.instance.websocket.users` to get the number of connected Websocket sessions, but also want to have metrics that can predict an increased demand in the future.


### Description
This PR adds additional metrics to get the number of upcoming and ending exams and exercises, grouped in buckets of 15, 30, 45, 60 and 120 minutes and additionally grouped by exercise type (for exercise metrics).
The number is available both as count and as `studentMultiplier` which takes the number of registered students into account.
Additionally, a value is exposed that only contains student that were active within the last 14 days.

The metric is available as `artemis.scheduled.{exams,exercises}.{due,release}.{count,student_multiplier,student_multiplier.active.14}`.
To only include active users, `student_multiplier.active.14` can be used instead of `student_multiplier`
The time range is available as tag `range` (with values `15`, `30`, `45`, `60`, `120`).
The exercise type is available as tag `exerciseType` (with values `QUIZ` , `TEXT`, `PROGRAMMING`, `MODELING`, `FILE_UPLOAD`).

So for example, the number of quiz exercises starting withing the next 30 minutes is available as `artemis_scheduled_exercises_release_count{application="Artemis",exerciseType="QUIZ",range="30",}` and the number of students registered for these exercises is available as `artemis_scheduled_exercises_release_student_multiplier{application="Artemis",exerciseType="QUIZ",range="30",}`.

Additionaly, metrics for public statistics (such as the number of courses, etc.) are exposed.
**Note**: As these metrics require somewhat large database queries, the performance should be monitored after the deployment.

### Steps for Testing
Prerequisites:
- Make sure you can reach the `ARTEMIS-HOST/management/prometheus` endpoint. You might have to add your ip address to the `spring.prometheus.monitoringIp` property (You can set it to `0.0.0.0/0,::1/0` to allow the access from any ip address). 
- The exercises and exams are not counted if they are created in a `Testing Course`, so make sure to use a "non-Testing Course"

1. Create multiple (e.g. 3) exercises with a start and release date within the next 2 hours
2. Make sure students are registered for the course
3. Make sure that the metrics `artemis_scheduled_exercises_release_count` return reasonable results and the `student_multiplier` metrics are also returned as expected
4. Repeat the step with exams and check the corresponding metrics


### Review Progress







#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage





| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| MetricsBean.java | 86% | ✅                           |
